File: PositronServer/internal/arena/transientArena.go
package arena

type TransientArena struct {
	buff     []byte
	writePtr uint32
}

func NewTransfientArena() *TransientArena {
	return &TransientArena{
		buff:     make([]byte, ALLOC_CHUNK),
		writePtr: 0,
	}
}

func (t *TransientArena) CloneFrom(externalBuffer []byte) {
	t.checkAndResizeAbsolute(uint32(len(externalBuffer)))
	t.writePtr = uint32(len(externalBuffer))
	copy(t.buff, externalBuffer)
}

func (t *TransientArena) Alloc(data []byte) uint32 {
	dataLen := uint32(len(data))

	t.checkAndResize(dataLen)

	copy(t.buff[t.writePtr:t.writePtr+dataLen], data)
	ptr := t.writePtr
	t.writePtr += dataLen

	return ptr
}

func (t *TransientArena) Read(ptr, len uint32) ([]byte, error) {
	if (ptr + len) > t.writePtr {
		return nil, ErrInvalidDescriptor
	}

	return t.buff[ptr:(ptr + len)], nil
}

func (t *TransientArena) ReadAll() []byte {
	return t.buff[:t.writePtr]
}

func (t *TransientArena) Flush() {
	clear(t.buff)
	t.writePtr = 0
}

func (t *TransientArena) GetActualSize() int {
	return int(t.writePtr)
}

func (t *TransientArena) checkAndResize(dataLen uint32) {
	buffLen := uint32(len(t.buff))

	if (dataLen + t.writePtr) <= buffLen {
		return
	}

	need := t.writePtr + dataLen
	t.checkAndResizeAbsolute(need)
}

func (t *TransientArena) checkAndResizeAbsolute(need uint32) {
	if uint32(len(t.buff)) >= need {
		return
	}

	for uint32(len(t.buff)) < need {
		t.buff = append(t.buff, make([]byte, ALLOC_CHUNK)...)
	}
}

File: PositronUnitySdk/PositronSdk/Assets/PositronSdk/Scripts/BytesArena/TransientArena.cs
using System;
using System.Text;

namespace Positron.BytesArena
{
    /// <summary>
    /// Custom memory manager for hot parts of system
    /// </summary>
    public sealed class TransientArena
    {
        private byte[] _buffer;
        private int _writePtr;

        /// <summary>
        /// Get`s all available data
        /// </summary>
        public ReadOnlySpan<byte> Data => _buffer.AsSpan(0, _writePtr);

        public TransientArena()
        {
            _buffer = new byte[64 * 1024]; //64 KB
        }

        /// <summary>
        /// Clones arena to internal buffer
        /// </summary>
        /// <param name="data">source arena</param>
        public void CloneFrom(ReadOnlyMemory<byte> data)
        {
            Flush();
            Alloc(data.Span, out int _);
        }

        /// <summary>
        /// Sequentically allocates memory in internal buffer without any headers
        /// </summary>
        /// <param name="data">raw bytes</param>
        /// <returns>pointer to internal buffer</returns>
        public int Alloc(ReadOnlySpan<byte> data, out int len)
        {
            int size = data.Length;
            int ptr = _writePtr;

            CheckSizeAndGrow(size);
            data.CopyTo(_buffer.AsSpan(_writePtr, size));
            _writePtr += size;

            len = size;
            return ptr;
        }

        /// <summary>
        /// Reads data from internal buffer
        /// </summary>
        /// <param name="ptr">offset (pointer) at buffer</param>
        /// <param name="len">lengths to read</param>
        /// <returns>Link to internals data (NOT A COPY!)</returns>
        /// <exception cref="IndexOutOfRangeException">Throwed when you try to read outside of physically allocated buffer</exception>
        public Span<byte> Read(uint ptr, uint len)
        {
            if ((ptr + len) > _buffer.Length)
            {
                throw new IndexOutOfRangeException($"Unable to read memory at {ptr} {len} {_buffer.Length}");
            }

            return _buffer.AsSpan((int)ptr, (int)len);
        }

        /// <summary>
        /// Reads data from internal buffer (as memory)
        /// </summary>
        /// <param name="ptr">offset (pointer) at buffer</param>
        /// <param name="len">lengths to read</param>
        /// <returns>Link to internals data (NOT A COPY!) as memory</returns>
        /// <exception cref="IndexOutOfRangeException">Throwed when you try to read outside of physically allocated buffer</exception>
        public Memory<byte> ReadAsMem(uint ptr, uint len)
        {
            if ((ptr + len) > _buffer.Length)
            {
                throw new IndexOutOfRangeException($"Unable to read memory at {ptr} {len} {_buffer.Length}");
            }

            return _buffer.AsMemory((int)ptr, (int)len);
        }

        /// <summary>
        /// All data GONE after that operation
        /// </summary>
        public void Flush() 
        {
            _buffer.AsSpan().Clear();
            _writePtr = 0;
        }

        private void CheckSizeAndGrow(int size)
        {
            if ((_writePtr + size) >= _buffer.Length)
            {
                byte[] doubled = new byte[_writePtr + (size * 2)];
                _buffer.AsSpan().CopyTo(doubled.AsSpan());
                _buffer = doubled;
            }
        }

        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();

            sb.AppendLine($"PTR: {_writePtr}");
            sb.AppendLine("DAT: ");

            for (int i = 0; i < _writePtr; i++)
            {
                sb.Append(' ');
                sb.Append(_buffer[i]);
                sb.Append(' ');
            }
            
            return sb.ToString();
        }
    }
}